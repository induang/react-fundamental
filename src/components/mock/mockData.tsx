export const mockedCourse = {
	id: 'testID',
	title: 'testTitle',
	description: 'testDescrip',
	creationDate: '09/10/2022',
	duration: 12,
	authors: ['testAuthorId_1', 'testAuthorId_2'],
};

export const mockedAuthor_1 = {
	id: 'testAuthorId_1',
	name: 'testAuthorName_1',
};

export const mockedAuthor_2 = {
	id: 'testAuthorId_2',
	name: 'testAuthorName_2',
};

export const mockedUser = {
	isAuth: true,
	name: 'TestName',
	email: 'test',
	token: 'test',
	role: 'test',
};
