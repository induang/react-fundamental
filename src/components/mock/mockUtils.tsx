import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import { ToolkitStore } from '@reduxjs/toolkit/dist/configureStore';

import { RootState } from '../../redux/store/store';
import userSlice from '../../redux/slices/userSlice';
import courseSlice from '../../redux/slices/courseSlice';
import authorsSlice from '../../redux/slices/authorsSlice';

import {
	mockedUser,
	mockedCourse,
	mockedAuthor_1,
	mockedAuthor_2,
} from './mockData';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: ToolkitStore;
}
export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {
			user: mockedUser,
			courses: {
				courses: [mockedCourse],
			},
			authors: {
				authors: [mockedAuthor_1, mockedAuthor_2],
			},
		},
		store = configureStore({
			reducer: { user: userSlice, courses: courseSlice, authors: authorsSlice },
			preloadedState,
		}),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return <Provider store={store}>{children}</Provider>;
	}
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
