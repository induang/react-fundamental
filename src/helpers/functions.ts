import moment from 'moment';

export const formatTime = (duration: number) => {
	if (Math.floor(duration / 60) < 10 && duration % 60 < 10)
		return `0${Math.floor(duration / 60)}:0${duration % 60}`;
	if (Math.floor(duration / 60) < 10)
		return `0${Math.floor(duration / 60)}:${duration % 60}`;
	if (duration % 60 < 10)
		return `${Math.floor(duration / 60)}:0${duration % 60}`;
	return `${Math.floor(duration / 60)}:${duration % 60}`;
};

export const formatDate = (creationDate: string) => {
	return moment(new Date(creationDate).getTime()).format('DD.MM.yyyy');
};

export const createTodayDate = () => {
	console.log('TODAY:', moment().format('MM/DD/yyyy'));
	return moment().format('MM/DD/yyyy');
};
