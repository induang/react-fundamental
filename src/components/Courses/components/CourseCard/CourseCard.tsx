import { useEffect } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Paper, Grid, Button, Box } from '@mui/material';
import EditRoundedIcon from '@mui/icons-material/EditRounded';
import DeleteIcon from '@mui/icons-material/Delete';

import { H1 } from '../../../../common/Title';
import { RootState } from '../../../../types';
import {
	Course,
	deleteCourseThunk,
} from '../../../../redux/slices/courseSlice';
import { Author, getAuthorsThunk } from '../../../../redux/slices/authorsSlice';
import { formatTime, formatDate } from '../../../../helpers/functions';
import {
	useAppSelector,
	useAppDispatch,
} from '../../../../redux/store/storeHook';

type CourseCardProps = {
	course: Course;
};

function CourseCard({ course }: CourseCardProps) {
	const user = useAppSelector((state: RootState) => state.user);
	const authors = useAppSelector((state: RootState) => state.authors.authors);
	const dispatch = useAppDispatch();

	const navigate = useNavigate();

	const handleDeleteCourseClick = (id: string) => {
		dispatch(deleteCourseThunk(id));
	};

	useEffect(() => {
		dispatch(getAuthorsThunk());
	}, []);

	return (
		<Paper elevation={2} className='my-1 p-1'>
			<Grid container spacing={2} className='p-5'>
				<Grid item md={7}>
					<H1 text={course.title} />
					<div className='w-11/12 indent-2'>{course.description}</div>
				</Grid>
				<Grid item md={5} className='flex-col justify-center'>
					<div id='author' className='w-80 truncate'>
						<b>Authors:&nbsp;&nbsp;</b>
						{course.authors.map(
							(id) =>
								authors.filter((author: Author) => author.id === id)[0]?.name
						) + ' '}
					</div>
					<div id='durations'>
						<b>Duration:&nbsp;&nbsp;</b>
						{`${formatTime(course.duration)} hours`}
					</div>
					<div id='created'>
						<b>Created:&nbsp;&nbsp;</b>
						{formatDate(course.creationDate)}
					</div>
					<div id='showCourseBtn'>
						<Box>
							<Grid
								container
								className='mt-5 justify-items-center items-center'
							>
								<Grid item>
									<Link to={`${course.id}`}>
										<Button variant='outlined' color='secondary'>
											Show Courses
										</Button>
									</Link>
								</Grid>
								{user.role === 'admin' ? (
									<Grid item>
										<Button
											color='secondary'
											onClick={() => navigate(`/courses/update/${course.id}`)}
										>
											<EditRoundedIcon />
										</Button>
									</Grid>
								) : (
									<></>
								)}
								{user.role === 'admin' ? (
									<Grid item>
										<Button
											color='secondary'
											onClick={() => handleDeleteCourseClick(course.id)}
										>
											<DeleteIcon />
										</Button>
									</Grid>
								) : (
									<></>
								)}
							</Grid>
						</Box>
					</div>
				</Grid>
			</Grid>
		</Paper>
	);
}

export default CourseCard;
