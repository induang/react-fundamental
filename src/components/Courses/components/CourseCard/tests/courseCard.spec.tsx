import { describe, it } from '@jest/globals';
import { BrowserRouter as Router } from 'react-router-dom';
import { renderWithProviders } from '../../../../mock/mockUtils';
import {
	mockedCourse,
	mockedAuthor_1,
	mockedAuthor_2,
} from '../../../../mock/mockData';
import CourseCard from '../CourseCard';
import '@testing-library/jest-dom';
import { formatDate, formatTime } from '../../../../../helpers/functions';

jest.mock('../../../../../services', () => {
	return {
		async fetchAddAuthor() {
			return {
				result: {
					mockedAuthor_1,
				},
			};
		},
		async fetchAuthorsData() {
			return {
				result: [mockedAuthor_1, mockedAuthor_2],
			};
		},
	};
});

window.alert = jest.fn(() => ({}));

describe('render courseCard', () => {
	it('render correctly', () => {
		const { container, debug } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		debug();
		expect(container.firstChild).toMatchSnapshot();
	});

	it('display course title', () => {
		const { getByRole } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		expect(getByRole('heading', { level: 1 })).toHaveTextContent(
			mockedCourse.title
		);
	});

	it('display course description', () => {
		const { getByText } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		expect(getByText(mockedCourse.description)).toBeTruthy();
	});

	it('display duration in the correct format', () => {
		const { queryByText } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		expect(queryByText(/duration/i)?.parentElement).toHaveTextContent(
			formatTime(mockedCourse.duration)
		);
	});
	it('display authors list', () => {
		expect.assertions(2);
		const { queryByText } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		let author = mockedAuthor_1.name;
		expect(queryByText(/Authors:/)?.parentElement).toHaveTextContent(
			new RegExp(author, 'i')
		);
		author = mockedAuthor_2.name;
		expect(queryByText(/Authors:/)?.parentElement).toHaveTextContent(
			new RegExp(author, 'i')
		);
	});

	it('display created date in the correct format', () => {
		const { queryByText } = renderWithProviders(
			<Router>
				<CourseCard {...{ course: mockedCourse }} />
			</Router>
		);
		expect(queryByText(/Created:/)?.parentElement).toHaveTextContent(
			formatDate(mockedCourse.creationDate)
		);
	});
});
