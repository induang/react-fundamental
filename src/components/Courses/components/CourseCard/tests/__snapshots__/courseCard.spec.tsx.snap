// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render courseCard render correctly 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 my-1 p-1 css-1mbunh-MuiPaper-root"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 p-5 css-mhc70k-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-7 css-3vt552-MuiGrid-root"
    >
      <h1
        class="text-4xl font-black"
      >
        testTitle
      </h1>
      <div
        class="w-11/12 indent-2"
      >
        testDescrip
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-5 flex-col justify-center css-yw6hac-MuiGrid-root"
    >
      <div
        class="w-80 truncate"
        id="author"
      >
        <b>
          Authors:  
        </b>
        testAuthorName_1,testAuthorName_2 
      </div>
      <div
        id="durations"
      >
        <b>
          Duration:  
        </b>
        00:12 hours
      </div>
      <div
        id="created"
      >
        <b>
          Created:  
        </b>
        10.09.2022
      </div>
      <div
        id="showCourseBtn"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiGrid-root MuiGrid-container mt-5 justify-items-center items-center css-11lq3yg-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <a
                href="/testID"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-wnpvqa-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Show Courses
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
